include/swift/AST/Types.h

```diff
@@ -2833,21 +2845,29 @@ class AnyFunctionType : public TypeBase {
 
     bool operator==(Param const &b) const {
       return (Label == b.Label &&
+              InternalLabel == b.InternalLabel &&
```

Comment:
It would definitely work, but I thought it’s unintuitive if the `==` operator silently ignores one of the properties altogether, that’s why I went with the dedicated method. 

Thinking about this some more, I think we should 
a) make `==` check for strict equality without canonicalizing types (or remove it entirely) and provide `equals` functions that check the various forms of equality a `Param` can take or
b) accept that `==` that the definition of `==` is a bit sloppy, embrace it, don’t check for `InternalLabel` and add a doc comment describing the catches.

I can see advantages in either approach (clearer semantics in (a) but more concise syntax in (b)). I think I tend to lean towards (a) but don’t have too strong feelings about it. Do you @rintaro?

Reply:
I'd suggest (b) in this PR. Here's the reasoning: Considering `Type::operator==` is `delete` we probably should delete `AnyFunctionType::Param::operator==` and make `isEqual` which compares "canonical" forms of the values. But, that also applies to `AnyFunctionType::Yield::operator==`. I don't think we should do it in this PR because `AnyFunctionType::Yield` is unrelated. Also we should not change the semantic meaning of `AnyFunctionType::Param::operator==` in this PR.

Reply:
Sounds good to me.

Reply:
After looking into it some more, I agree that ignoring the internal name in `operator==` is *absolutely* the right choice. I wasn’t aware that we are *always* using canonical types in `operator==` for types and that it’s rather more difficult to compare the sugared types. As such, ignoring the internal name in `Param::operator==` follows the current paradigm.